dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java

```diff
@@ -2685,34 +2698,169 @@ private int getJobInstanceTableVersion(PersistenceServiceUnit psu) throws Except
+                                                                        JobInstanceEntityV3.class);
+            query.setParameter("jobinstanceid", jobInstanceID);
+            query.setParameter("groups", subjectGroups);
+
+            boolean notFound = query.setMaxResults(1).getResultList().isEmpty();
+            if (notFound) //the query did not find anything - the resulting list is empty
+                return false;
+            else
+                return true; //the resulting list contains something that satisfied the query - we don't care what it is, just that there is at least one record present
+
+        } finally {
+            em.close();
+        }
+    }
+
+    /** {@inheritDoc} */
+//    @Override
+//    public Collection<? extends WSJobInstance> getJobInstancesWithGroupSecurity(WSSearchObject wsso, int page, int pageSize) {
+//        // TODO Auto-generated method stub
+//        return null;
+//    }
+
+    /** {@inheritDoc} */
+    @Override
+    public List<JobInstanceEntity> getJobInstancesForSubjectGroupNames(List<String> listOfGroupsForSubject, String subject) {
+        EntityManager em = getPsu().createEntityManager();
+        List<JobInstanceEntity> jobInstances = null;
+
+        try {
+            TypedQuery<JobInstanceEntityV3> query = em.createNamedQuery(JobInstanceEntityV3.GET_JOBINSTANCES_FIND_BY_SUBMITTER_OR_GROUPACCESS_QUERY,
+                                                                        JobInstanceEntityV3.class);
+            query.setParameter("groups", listOfGroupsForSubject);
+            //query.setParameter("submitterName", subject);
+            List<JobInstanceEntityV3> rows = query.getResultList();
+
+            if (rows.isEmpty()) {
+                jobInstances = new ArrayList<JobInstanceEntity>();
+
+            } else {
+                jobInstances = new ArrayList<JobInstanceEntity>(rows);
+            }
+
+            return jobInstances;
+        } finally {
+            em.close();
+        }
+    }
+
+    @Override
+    public JobInstanceEntity updateJobInstanceWithGroupNames(final long jobInstanceID, final Set<String> groupNames) {
+
+        EntityManager em = getPsu().createEntityManager();
+
+        try {
+            return new TranRequest<JobInstanceEntityV3>(em) {
+                @Override
+                public JobInstanceEntityV3 call() {
+
+                    JobInstanceEntityV3 instance = entityMgr.find(JobInstanceEntityV3.class, jobInstanceID);
+
+                    if (instance == null) {
+                        throw new NoSuchJobInstanceException("No job instance found for id = " + jobInstanceID);
+                    }
+
+                    instance.setGroupName(groupNames);
+
+                    entityMgr.merge(instance);
+                    return instance;
+                }
+            }.runInNewOrExistingGlobalTran();
+        } finally {
+            em.close();
+        }
+    }
+
 }
```

Comment:
I know this is a huge class already, but can we try to keep more to the sort-of-pattern we have going for ordering the methods?

Reply:
class is organized by object type (instance methods then execution methods etc)